### PR TITLE
AG-17366-full-colmodel-rewrite

### PR DESCRIPTION
**Pull request by SalvatorePreviti**
Complete rewrite of ColumnModel to address bugs, performance, and unneded refreshes in many scenarios

FIX AG-17366

Bugs fixed:

FIX AG-16668
FIX AG-14362

implements also parts of AG-13046 and AG-13147

- Fixes O(n) lookup for getCol by AgColumn with a single optimized record O(1) lookup, faster than map for small list of columns
- Solves the prototype conflict in column model using Object.create(null) instead of {} as Records and add tests for it, so __proto__, constructor, toString and other Object.prototype methods risk there does not exist anymore and can be a colId
- Standardise the way we build columns, replacing many O(n*n) or worse with O(n) and O(1) operations, keeping the grid fast as the number of columns increase, and keeping it faster also for small number of columns.
- Full reuse of AgColumn for services and pivot and grouping and all the service columns like row number or selection - column groups, column span, provided column groups included, the tree is fully diffed correctly.
- Less refresh when colDef don't change through the new _mergedEqual utility - we check for structural equality before 
- Removal of manually custom managed logic for columns list in CalculatedColSvc as now the ColModel can handle them fine with the new structure that allows to define the new column tree shape with O(1) operations without array manipulations - this also ensures calculated columns are fully managed with all edge cases for padding, row span, col span, and everything that column model have to handle
- Fixes bugs and improves the performance of row span
- Fixes ~100 tests / bugs that in latest are marked as skip and new tests in this PR 
- Fixes a bug with hierarchical column counting wrongly the Quarters - fixed and added tests 

In the previous PRs we merged 70000 lines of tests in ad-hoc test PRs and feature PRs, for example calculated columns. This adds a good safety net against regression. The main changed files have a 100% code coverage, measured with vitest code coverage, and most importantly behavioural coverage.

## Changes in grid codebase (excluding tests):

files:         10 added, 5 removed, 142 changed (157 total)
lines added:   +6884
lines removed: -7179
net delta:     -295

Note that the changes are concentrated mostly in few files, the number of files changed are because the internal APIs changed and so those files needed fixes.

Add more tests and benchmarks

## Benchmarks:

  ↑   306.15x faster  [getvalue] getDataValue by Column object (last of 100 cols)  (55.31 → 16,932 ops/s)
  ↑     8.87x faster  [column-update] getColumnState 50 flat cols  (325,310 → 2,886,801 ops/s)
  ↑     7.23x faster  [column-refresh] setColumnDefs unchanged with maintainColumnOrder (fast path)  (133.7 → 966.5 ops/s)
  ↑     6.73x faster  [column-refresh] setColumnDefs 20 flat cols (no-op fast path, same ref)  (136.7 → 920.5 ops/s)
  ↑     6.63x faster  [column-refresh] setColumnDefs 20 flat cols (no-op fast path, equal-shape new ref)  (141.5 → 937.9 ops/s)
  ↑     3.92x faster  [column-refresh] maintainColumnOrder grouped churn (10 groups, 10 new cols/refresh)  (42.44 → 166.4 ops/s)
  ↑     2.90x faster  [column-update] resetColumnState 50 cols  (495.1 → 1437.0 ops/s)
  ↑     1.92x faster  [column-refresh] toggle rowNumbers on/off (20 cols)  (205.9 → 396.0 ops/s)
  ↑     1.62x faster  [getvalue] getDataValue by string (last of 100 cols)  (10,907 → 17,697 ops/s)
  ↑     1.55x faster  [column-refresh] addRowGroupColumns / removeRowGroupColumns (20 cols)  (142.1 → 220.6 ops/s)
  ↑     1.50x faster  [column-update] applyColumnState multi-sort 6 of 50 cols  (612.9 → 917.5 ops/s)
  ↑     1.46x faster  [column-update] applyColumnState restore saved state 50 flat cols  (913.9 → 1338.8 ops/s)
  ↑     1.33x faster  [column-update] applyColumnState toggle pivot (pivotMode) 20 cols  (474.1 → 628.2 ops/s)
  ↑     1.32x faster  [pivot-aggregation] transaction update (500 rows)  (293.3 → 387.0 ops/s)
  ↑     1.31x faster  [column-update] applyColumnState reverse/forward order 8 groups × 5 cols (applyOrder)  (717.8 → 939.6 ops/s)
  ↑     1.30x faster  [column-update] applyColumnState reverse/forward order 50 cols (applyOrder)  (725.7 → 942.1 ops/s)
  ↑     1.26x faster  [pivot-aggregation] transaction update (500 rows)  (302.2 → 380.3 ops/s)
  ↑     1.24x faster  [column-update] applyColumnState toggle aggFunc on value col 20 cols  (614.2 → 762.5 ops/s)
  ↑     1.22x faster  [column-update] applyColumnState toggle pinned 5 of 50 cols  (375.3 → 459.6 ops/s)
  ↑     1.22x faster  [column-refresh] move column in 100 flat cols (pure visibleCols.refresh)  (520.2 → 634.7 ops/s)
  ↑     1.20x faster  [column-update] applyColumnState toggle rowGroup (auto col churn) 20 cols  (358.4 → 431.3 ops/s)
  ↑     1.18x faster  [column-update] applyColumnState set width 50 cols  (482.8 → 570.0 ops/s)
  ↑     1.18x faster  [pivot-aggregation] immutable update (500 rows)  (236.2 → 278.6 ops/s)
  ↑     1.18x faster  [column-refresh] move column in 100 grouped cols (pure visibleCols.refresh)  (414.0 → 488.3 ops/s)
  ↑     1.17x faster  [column-refresh] toggle rowSelection on/off (20 cols)  (216.7 → 254.5 ops/s)
  ↑     1.17x faster  [pivot-aggregation] immutable update (500 rows)  (236.6 → 276.7 ops/s)
  ↑     1.12x faster  [getvalue] getCellValue by string (last of 100 cols)  (11,478 → 12,910 ops/s)
  ↑     1.12x faster  [pivot-aggregation] full refresh  (38.76 → 43.26 ops/s)
  ↑     1.10x faster  [pivot-aggregation] transaction update (500 rows)  (60.48 → 66.56 ops/s)
  ↑     1.10x faster  [pivot-aggregation] full refresh  (37.77 → 41.48 ops/s)
  ↑     1.09x faster  [pivot-aggregation] transaction update (500 rows)  (59.98 → 65.28 ops/s)
  ↑     1.08x faster  [column-refresh] setColumnDefs 10 flat cols (alternating defs)  (202.6 → 219.8 ops/s)
  ↑     1.08x faster  [column-update] applyColumnState set flex 50 cols  (1011.2 → 1091.6 ops/s)
  ↑     1.08x faster  [flat-sorting] sort 30000 rows  (27.20 → 29.32 ops/s)
  ↑     1.07x faster  [delta-sort-transaction] applyTransaction (deltaSort: false) - 1 update  (7.028 → 7.554 ops/s)
  ↑     1.07x faster  [formulas] warm: re-read sum/product/branch for 3000 rows (no invalidation)  (812.9 → 869.0 ops/s)

## Potential follow up for calculated columns

At the moment changing them causes a double visibility refresh and some double refresh events, when using applyColumnState or resetColumnState as we need to first ensure the tree is rebuilt correctly

### CODE REVIEW

File: packages/ag-grid-enterprise/src/columns/baseColsService.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+     *  by the imperative API; the state/agg-func paths drive those themselves, so they pass false. */
+    protected setColActive(col: AgColumn, active: boolean, source: ColumnEventType, withSideEffects = false): boolean {
+        if (this.isColActive(col) === active) {
+            return false;
+        }
+        if (active) {
+            this.addActiveCol(col);
+            this.setColFlag(col, active, source);
+            this.activateVirtuals(col, source);
+        } else {
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply packages/ag-grid-enterprise/src/columns/baseColsService.ts:85]: False positive - this is the intended behaviour, added more tests to really lock down the behaviour as is and document it

File: packages/ag-grid-community/src/columns/columnModel.ts [hunk 2]
```diff
@@ -104,633 +175,453 @@ export class ColumnModel extends BeanStub implements NamedBean {
...
         } = beans;
-        // only need to dispatch before/after events if updating columns, never if setting columns for first time
-        const dispatchEventsFunc = this.colDefCols ? _compareColumnStatesAndDispatchEvents(beans, source) : undefined;
 
-        // always invalidate cache on changing columns, as the column id's for the new columns
-        // could overlap with the old id's, so the cache would return old values for new columns.
+        // only dispatch before/after events when updating columns, not on first set
+        const colDefs = this.colDefs;
+        const stateChanges = colDefs ? captureColumnStateChanges(beans) : undefined;
+        // new ids may collide with old ids, so cached values would be wrong for new cols
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply packages/ag-grid-community/src/columns/columnModel.ts]: edge case not reproducible by API but improved and added tests

**Review by StephenCooper — APPROVED** [review]
Happy to proceed.
